File: web/scripts/extract-wordle-bank.mjs
import fs from 'fs';
import path from 'path';
import { createRequire } from 'module';
import { fileURLToPath } from 'url';

const require = createRequire(import.meta.url);
const { PDFParse } = require('pdf-parse');

const __dirname = path.dirname(fileURLToPath(import.meta.url));
const downloads = path.join(
  process.env.USERPROFILE,
  'OneDrive - Tahoma School District',
  'Downloads',
);

const files = [
  { key: 'fry', file: 'fry_complete_1000.pdf' },
  { key: 'dolch', file: 'Copy of Dolch-Words-Flash-Cards.pdf' },
  { key: 'heart', file: 'Copy of Heart-Words.pdf' },
  { key: 'irregular', file: 'Copy of Irregular-Words-Flash-Cards.pdf' },
];

function extractWords(text) {
  const lines = text.split(/\r?\n/).map((l) => l.trim()).filter(Boolean);
  const words = [];
  for (const line of lines) {
    if (/^--\s*\d+\s+of\s+\d+\s*--$/i.test(line)) continue;
    if (/^list\s+\d+$/i.test(line)) continue;
    if (/fry words/i.test(line)) continue;
    if (/copyright/i.test(line)) continue;
    if (/k12reader/i.test(line)) continue;
    if (/dolch words/i.test(line)) continue;
    if (/flash cards/i.test(line)) continue;
    if (/pre-?primer/i.test(line)) continue;
    if (/^(primer|first grade|second grade|third grade)$/i.test(line)) continue;
    if (/irregular/i.test(line) && line.length > 20) continue;
    if (/heart.?words/i.test(line) && line.length > 15) continue;
    if (line.length > 24) continue;
    if (!/^[A-Za-z][A-Za-z']*$/.test(line)) continue;
    const clean = line.toLowerCase().replace(/[^a-z]/g, '');
    if (clean.length < 2 || clean.length > 8) continue;
    words.push(clean);
  }
  return words;
}

async function pdfText(filePath) {
  const parser = new PDFParse({ data: fs.readFileSync(filePath) });
  try {
    const result = await parser.getText();
    return result?.text || '';
  } finally {
    await parser.destroy?.();
  }
}

const bySource = {};
const all = new Set();

for (const { key, file } of files) {
  const text = await pdfText(path.join(downloads, file));
  const words = extractWords(text);
  bySource[key] = [...new Set(words)].sort();
  for (const w of bySource[key]) all.add(w);
}

const byLength = {};
for (let n = 2; n <= 8; n++) byLength[n] = [];
for (const w of [...all].sort()) {
  if (byLength[w.length]) byLength[w.length].push(w);
}

const outDir = path.join(__dirname, '..', 'src', 'apps', 'games', 'wordle');
fs.mkdirSync(outDir, { recursive: true });

const meta = {
  generatedAt: new Date().toISOString(),
  sources: Object.fromEntries(Object.entries(bySource).map(([k, v]) => [k, v.length])),
  totalUnique: all.size,
  byLengthCounts: Object.fromEntries(
    Object.entries(byLength).map(([k, v]) => [k, v.length]),
  ),
};

const heartWords = [
  ...new Set([...(bySource.heart || []), ...(bySource.irregular || [])]),
].sort();
const highFrequency = [
  ...new Set([...(bySource.fry || []), ...(bySource.dolch || [])]),
].sort();

const fileBody = `/** Auto-generated high-frequency + heart/irregular word bank for Wordle (2–8 letters). */
export const WORD_BANK_META = ${JSON.stringify(meta, null, 2)};

/** Words grouped by length (lowercase, a–z only). */
export const WORDS_BY_LENGTH = ${JSON.stringify(byLength, null, 2)};

/** Heart / irregular words (subset). */
export const HEART_WORDS = new Set(${JSON.stringify(heartWords)});

/** Fry + Dolch high-frequency words. */
export const HIGH_FREQUENCY_WORDS = new Set(${JSON.stringify(highFrequency)});

export function wordsForLength(len) {
  return WORDS_BY_LENGTH[len] || [];
}

export function pickRandomWord(len) {
  const list = wordsForLength(len);
  if (!list.length) return null;
  return list[Math.floor(Math.random() * list.length)];
}
`;

fs.writeFileSync(path.join(outDir, 'wordBank.js'), fileBody);
console.log(JSON.stringify(meta, null, 2));
console.log('sample5', byLength[5]?.slice(0, 20));

File: web/scripts/extract-wordle-bank.py
#!/usr/bin/env python3
"""Extract Fry / Dolch / UFLI Wordle lists from PDFs into wordLists.js."""

from __future__ import annotations

import json
import re
import unicodedata
from pathlib import Path

import fitz

DOWNLOADS = Path.home() / "Downloads"
OUT = (
    Path(__file__).resolve().parent.parent
    / "src"
    / "apps"
    / "games"
    / "wordle"
    / "wordLists.js"
)

SOURCES = [
    (
        "fry",
        "Fry 1000",
        "Fry high-frequency words (first 1,000).",
        DOWNLOADS / "fry_complete_1000.pdf",
    ),
    (
        "dolch",
        "Dolch",
        "Dolch sight words (pre-primer through third grade).",
        DOWNLOADS / "Dolch Words Slides.pdf",
    ),
    (
        "heart",
        "UFLI Heart Words",
        "UFLI irregular heart words.",
        DOWNLOADS / "UFLI Heart Words.pdf",
    ),
    (
        "irregular",
        "UFLI Irregular Words",
        "UFLI irregular words slides.",
        DOWNLOADS / "UFLI Irregular Words.pdf",
    ),
]

SKIP_LINE = re.compile(
    r"(copyright|k12reader|www\.|worksheet|free to duplicate|free for educational|"
    r"fry words|list\s+\d+|slides|irregular|heart.?words|dolch|"
    r"pre-?primer|^primer$|first grade|second grade|third grade|"
    r"nouns|verbs|adjectives|sight words)",
    re.I,
)


def normalize_word(s: str) -> str:
    s = unicodedata.normalize("NFKC", s).strip()
    s = (
        s.replace("ﬁ", "fi")
        .replace("ﬂ", "fl")
        .replace("ﬀ", "ff")
        .replace("ﬃ", "ffi")
        .replace("ﬄ", "ffl")
        .replace("’", "'")
        .replace("‘", "'")
    )
    return s


def extract_words(pdf_path: Path) -> list[str]:
    doc = fitz.open(pdf_path)
    words: list[str] = []
    for page in doc:
        for raw in page.get_text("text").splitlines():
            line = normalize_word(raw)
            if not line:
                continue
            if SKIP_LINE.search(line) and not re.fullmatch(r"[A-Za-z']+", line):
                continue
            if SKIP_LINE.fullmatch(line):
                continue
            if not re.fullmatch(r"[A-Za-z']+", line):
                continue
            if "'" in line:
                continue
            clean = line.lower()
            if len(clean) < 2 or len(clean) > 8:
                continue
            if not re.fullmatch(r"[a-z]+", clean):
                continue
            words.append(clean)
    return sorted(set(words))


def main() -> None:
    lists = []
    for list_id, name, description, path in SOURCES:
        words = extract_words(path)
        lists.append(
            {
                "id": list_id,
                "name": name,
                "description": description,
                "words": words,
            }
        )
        print(f"{list_id}: {len(words)} words from {path.name}")

    lines = [
        "/** Classroom Wordle lists extracted from Fry, Dolch, and UFLI PDFs (2–8 letters, a–z only). */",
        "export const WORD_LISTS = [",
    ]
    for lst in lists:
        lines.append("  {")
        lines.append(f"    id: {json.dumps(lst['id'])},")
        lines.append(f"    name: {json.dumps(lst['name'])},")
        lines.append(f"    description: {json.dumps(lst['description'])},")
        words_js = ", ".join(json.dumps(w) for w in lst["words"])
        lines.append(f"    words: [{words_js}],")
        lines.append("  },")
    lines.append("];")
    lines.append("")
    lines.append("export const WORD_LIST_IDS = WORD_LISTS.map((list) => list.id);")
    lines.append("")
    OUT.write_text("\n".join(lines) + "\n")
    print("wrote", OUT)


if __name__ == "__main__":
    main()
